pkg/reconciler/topic/topic.go

```diff
@@ -0,0 +1,420 @@
+	servinglisters "github.com/knative/serving/pkg/client/listers/serving/v1beta1"
+	"go.uber.org/zap"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+
+	"github.com/GoogleCloudPlatform/cloud-run-events/pkg/apis/pubsub/v1alpha1"
+	listers "github.com/GoogleCloudPlatform/cloud-run-events/pkg/client/listers/pubsub/v1alpha1"
+	"github.com/GoogleCloudPlatform/cloud-run-events/pkg/reconciler/pubsub"
+	"github.com/GoogleCloudPlatform/cloud-run-events/pkg/reconciler/topic/resources"
+)
+
+const (
+	// ReconcilerName is the name of the reconciler
+	ReconcilerName = "Topics"
+
+	finalizerName = controllerAgentName
+)
+
+// Reconciler implements controller.Reconciler for Topic resources.
+type Reconciler struct {
+	*pubsub.PubSubBase
+
+	// listers index properties about resources
+	topicLister   listers.TopicLister
+	serviceLister servinglisters.ServiceLister
+
+	publisherImage string
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Service resource
+// with the current status of the resource.
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		c.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+	logger := logging.FromContext(ctx)
+
+	// Get the Topic resource with this namespace/name
+	original, err := c.topicLister.Topics(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logger.Errorf("service %q in work queue no longer exists", key)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	Topic := original.DeepCopy()
+
+	// Reconcile this copy of the Topic and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	var reconcileErr = c.reconcile(ctx, Topic)
+
+	// If no error is returned, mark the observed generation.
+	if reconcileErr == nil {
+		Topic.Status.ObservedGeneration = Topic.Generation
+	}
+
+	if equality.Semantic.DeepEqual(original.Finalizers, Topic.Finalizers) {
+		// If we didn't change finalizers then don't call updateFinalizers.
+
+	} else if _, updated, fErr := c.updateFinalizers(ctx, Topic); fErr != nil {
+		logger.Warnw("Failed to update Topic finalizers", zap.Error(fErr))
+		c.Recorder.Eventf(Topic, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update finalizers for Topic %q: %v", Topic.Name, fErr)
+		return fErr
+	} else if updated {
+		// There was a difference and updateFinalizers said it updated and did not return an error.
+		c.Recorder.Eventf(Topic, corev1.EventTypeNormal, "Updated", "Updated Topic %q finalizers", Topic.GetName())
+	}
```

Comment:
This if-else sequence is not clear to me. I think it would be clearer as two nested ifs:

```go
if !equality.Semantic.DeepEqual(original.Finalizers, Topic.Finalizers) {
  _, updated, fErr := c.updateFinalizers(ctx, Topic)
  if ferr != nil {
    // handle error
    // if this were a helper function, you could return here and avoid the else
  } else if updated {
    // handle success
  }
}
```

Reply:
I am keeping formatting with sample-controller:

https://github.com/knative/sample-controller/blob/master/pkg/reconciler/addressableservice/addressableservice.go#L92-L106